### PR TITLE
fix(execution): Canister snapshots should include Wasm globals

### PR DESCRIPTION
**Pull request by luc-blaeser**
Currently, the canister snapshots do not cover the state of Wasm global variables, such that in realistic canister programs with mutable global variables, the state is not properly recovered when a snapshot is loaded.

This fix adds the Wasm globals to the snapshot state, similar to the execution state. (Side note: All Wasm variables are exported by the Wasm instrumentation.)

### CODE REVIEW

File: rs/replicated_state/src/canister_snapshots.rs [hunk 1]
```diff
@@ -284,6 +284,10 @@ pub struct ExecutionStateSnapshot {
     /// The raw canister module.
     #[validate_eq(Ignore)]
     pub wasm_binary: CanisterModule,
+    /// The Wasm global variables.
+    /// Note: The hypervisor instrumentations exports all global variables,
+    /// including originally internal global variables.
+    pub exported_globals: Vec<Global>,
```
Comment (schneiderstefan) [inline rs/replicated_state/src/canister_snapshots.rs:290]:
```suggestion
    #[validate_eq(Ignore)]
    pub exported_globals: Vec<Global>,
```

This field should be treated the same as the corresponding field in the canister for the purpose of validation during checkpoints.

**Review by schneiderstefan — COMMENTED** [review]
I think the same problem exists for the fields `ExecutionState::exports` and `ExecutionState::metadata`.

**Comment by berestovskyy** [comment]
> I think the same problem exists for the fields ExecutionState::exports and ExecutionState::metadata.

Why do you think so?

**Comment by berestovskyy** [comment]
> Aren't they also extracted from the wasm and should therefore be restored when the wasm is restored?

I believe the exports and metadata should be restored from the restored Wasm module. And I guess that's why we missed globals...

**Comment by schneiderstefan** [comment]
> I believe the exports and metadata should be restored from the restored Wasm module. And I guess that's why we missed globals...

I got it wrong then and it's all good.

File: rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs [hunk 2]
```diff
@@ -1714,3 +1714,91 @@ fn load_canister_snapshot_charges_canister_cycles() {
...
+        (export "memory" (memory $memory))
+        (global (export "counter") (mut i32) (i32.const 0))
+        (export "canister_query read_global" (func $read_global))
+        (export "canister_update increase_global" (func $increase_global))
+      )"#;
+    let wasm = wat::parse_str(wat).unwrap();
+
+    const CYCLES: Cycles = Cycles::new(1_000_000_000_000_000);
+    let own_subnet = subnet_test_id(1);
+    let caller_canister = canister_test_id(1);
```
Comment (berestovskyy) [inline rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs:1757]: nit: I'm not sure if the specific subnet and caller, as well as the reserved cycles, make a difference for this test. But we can clean them up later...

**Review by berestovskyy — APPROVED** [review]
Thanks a lot, @luc-blaeser!

File: rs/replicated_state/src/canister_snapshots.rs (hunk 1 above)
Comment (luc-blaeser) [reply rs/replicated_state/src/canister_snapshots.rs:291]: Thank you, I adjusted the code analogous to the canister state. I assume it is okay to skip this during checkpoint validation?

**Review by schneiderstefan — APPROVED** [review]

File: rs/replicated_state/src/canister_snapshots.rs (hunk 1 above)
Comment (schneiderstefan) [reply rs/replicated_state/src/canister_snapshots.rs:291]: Yeah it's one of those fields where we weren't sure if it's expensive to compare, so we set it to ignore for now. We might change that in the future, but either way canister states and snapshots should treat it the same.

File: rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs (hunk 2 above)
Comment (luc-blaeser) [reply rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs]: Thank you! I simplified the test.

**Review by dsarlis — APPROVED** [review]

**Pull request merged**